### PR TITLE
[feature] Port of Densify Geometries by Count algorithm to C++

### PR DESCRIPTION
**Pull request by root676**
## Description
This PR contains the port of the Densify Geometries by Count algorithm to a native C++ algorithm. Overall the performance of the algorithm is improved and a dynamic parameter for the number of vertices to add is implemented. Other than that, no other functionality has been added to the algorithm. 

I am looking forward to any reviews/comments! 

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Review by Gustry — COMMENTED** [review]
Nice, thanks.
It seems, you didn't remove old python files and the previous help description in the YML file.

**Comment by root676** [comment]
@Gustry thanks for the hint! I just deleted the old algorithm and the helpstrings.

File: src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+QString QgsDensifyGeometriesByCountAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorgeometry" );
+}
+
+QString QgsDensifyGeometriesByCountAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm takes a polygon or line layer"
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp]: Be careful of missing trailing spaces here -- this string is missing spaces between "layerand", "geometrieshave", etc.

Also would be good to add `\n\n` in places where the original help had a paragraph break

File: src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+void QgsDensifyGeometriesByCountAlgorithm::initParameters( const QVariantMap &configuration )
+{
+  Q_UNUSED( configuration )
+  std::unique_ptr<QgsProcessingParameterNumber> verticesCnt = qgis::make_unique<QgsProcessingParameterNumber>( QStringLiteral( "VERTICES" ),
+      QObject::tr( "Number of vertices to add" ),
+      QgsProcessingParameterNumber::Integer,
+      1, false, 1, 10000000 );
+  verticesCnt->setIsDynamic( true );
+  verticesCnt->setDynamicPropertyDefinition( QgsPropertyDefinition( QStringLiteral( "VerticesCount" ), QObject::tr( "Vertices count" ), QgsPropertyDefinition::IntegerPositiveGreaterZero ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp]:
```suggestion
  verticesCnt->setDynamicPropertyDefinition( QgsPropertyDefinition( QStringLiteral( "VerticesCount" ), QObject::tr( "Vertices count" ), QgsPropertyDefinition::IntegerPositive ) );
```

I think we should allow zero counts here -- e.g. if you only want to densify lines greater than a certain threshold

File: src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+QgsFeatureList QgsDensifyGeometriesByCountAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( context )
+  Q_UNUSED( feedback )
+  QgsFeature densifiedFeature = feature;
+
+  int verticesCnt = mVerticesCnt;
+  if ( mDynamicVerticesCnt )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp]: Very minor optimisation, but this logic could be moved into the `if ( feature.hasGeometry() )` branch -- it's unused when the feature has no geometry

File: src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    verticesCnt = mVerticesCntProperty.valueAsInt( context.expressionContext(), verticesCnt );
+
+  if ( feature.hasGeometry() )
+    densifiedFeature.setGeometry( feature.geometry().densifyByCount( verticesCnt ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmdensifygeometriesbycount.cpp]: I think we should put a safety check here of `if ( verticesCnt > 0 )`. It's possible a data defined parameter value may return a negative count... and I'm unsure if densifyByCount will handle that gracefully

**Review by nyalldawson — COMMENTED** [review]
Looks great!!

**Comment by root676** [comment]
Thank you @nyalldawson for reviewing this port and your feedback! I added the suggested changes.

**Pull request merged**